cpu/lpc2387/periph/i2c.c

```diff
@@ -0,0 +1,477 @@
+
+    ctx[dev].buf_cur = 0;
+    ctx[dev].cur     = ctx[dev].buf[0];
+    ctx[dev].end     = ctx[dev].buf_end[0];
+}
+
+int i2c_read_bytes(i2c_t dev, uint16_t addr,
+                   void *data, size_t len, uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+    lpc23xx_i2c_t *i2c = i2c_config[dev].dev;
+
+    /* Check for wrong arguments given */
+    if (data == NULL || len == 0) {
+        return -EINVAL;
+    }
+
+    /* TODO: 10 bit addresses */
+    if (flags) {
+        return -ENOTSUP;
+    }
+
+    _init_buffer(dev, 0, 1 | (addr << 1), (void*)data, len);
+
+    /* set Start flag */
+    i2c->CONSET = I2CONSET_STA;
+
+    mutex_lock(&ctx[dev].tx_done);
+    return ctx[dev].res;
+}
+
+int i2c_write_bytes(i2c_t dev, uint16_t addr, const void *data, size_t len,
+                    uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+    lpc23xx_i2c_t *i2c = i2c_config[dev].dev;
+
+    /* Check for wrong arguments given */
+    if (data == NULL || len == 0) {
+        return -EINVAL;
+    }
+
+    /* TODO: 10 bit addresses */
+    if (flags) {
+        return -ENOTSUP;
+    }
+
+    _init_buffer(dev, 0, addr << 1, (void*)data, len);
+
+    /* set Start flag */
+    i2c->CONSET = I2CONSET_STA;
+
+    mutex_lock(&ctx[dev].tx_done);
+    return ctx[dev].res;
+}
+
+int i2c_read_regs(i2c_t dev, uint16_t addr, uint16_t reg,
+                  void *data, size_t len, uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+    lpc23xx_i2c_t *i2c = i2c_config[dev].dev;
+
+    /* Check for wrong arguments given */
+    if (data == NULL || len == 0) {
+        return -EINVAL;
+    }
+
+    /* TODO: 10 bit addresses */
+    if (flags & ~I2C_REG16) {
+        return -ENOTSUP;
+    }
+
+    /* Handle endianness of register if 16 bit */
+    if (flags & I2C_REG16) {
+        reg = htons(reg); /* Make sure register is in big-endian on I2C bus */
+    }
+
+    _init_buffer(dev, 0, addr << 1, (void*)&reg, (flags & I2C_REG16) ? 2 : 1);
+    _init_buffer(dev, 1, 1 | (addr << 1), (void*)data, len);
+
+    /* set Start flag */
+    i2c->CONSET = I2CONSET_STA;
+
+    mutex_lock(&ctx[dev].tx_done);
+    return ctx[dev].res;
+}
+
+int i2c_write_regs(i2c_t dev, uint16_t addr, uint16_t reg,
+                   const void *data, size_t len, uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+    lpc23xx_i2c_t *i2c = i2c_config[dev].dev;
+
+    /* Check for wrong arguments given */
+    if (data == NULL || len == 0) {
+        return -EINVAL;
+    }
+
+    /* TODO: 10 bit addresses */
+    if (flags & ~I2C_REG16) {
+        return -ENOTSUP;
+    }
+
+    /* Handle endianness of register if 16 bit */
+    if (flags & I2C_REG16) {
+        reg = htons(reg); /* Make sure register is in big-endian on I2C bus */
+    }
+
+    _init_buffer(dev, 0, addr << 1, (void*)&reg, (flags & I2C_REG16) ? 2 : 1);
+    _init_buffer(dev, 1, addr << 1, (void*)data, len);
+
+    /* set Start flag */
+    i2c->CONSET = I2CONSET_STA;
+
+    mutex_lock(&ctx[dev].tx_done);
+    return ctx[dev].res;
+}
```

Comment:
Wouldn't is simplify the whole code to just use the implementations in `drivers/periph_common/i2c.c` and safe some RAM as only one buffer would be needed?

Reply:
I've spend too much time trying that.
The problem was that the device's internal state machine kept advancing unless I would issue a STOP condition.

From the data sheet I've gathered that the state machine would only advance after clearing the `I2CONCLR_SIC` bit, but that didn't seem to be the case.
The device would keep generating interrupts (and reading / writing bytes) until I set `I2CONSET_STO`.

Reply:
OK, I see